pkg/runtime/compstore/subscriptions.go

```diff
@@ -14,95 +14,195 @@ limitations under the License.
 }
 
-func (c *ComponentStore) DeleteTopicRoute(name string) {
+func (c *ComponentStore) AddDeclarativeSubscription(comp *subapi.Subscription, sub rtpubsub.Subscription) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	delete(c.topicRoutes, name)
-}
+	for i, existing := range c.subscriptions.declarativesList {
+		if existing == comp.Name {
+			c.subscriptions.declarativesList = append(c.subscriptions.declarativesList[:i], c.subscriptions.declarativesList[i+1:]...)
+			break
+		}
+	}
 
-func (c *ComponentStore) GetTopicRoutes() map[string]TopicRoutes {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-	return c.topicRoutes
+	c.subscriptions.declaratives[comp.Name] = &DeclarativeSubscription{
+		Comp:         comp,
+		Subscription: sub,
+	}
+	c.subscriptions.declarativesList = append(c.subscriptions.declarativesList, comp.Name)
 }
 
-func (c *ComponentStore) SetSubscriptions(subscriptions []rtpubsub.Subscription) {
+func (c *ComponentStore) AddStreamSubscription(comp *subapi.Subscription) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.subscriptions = subscriptions
+	c.subscriptions.streams[comp.Name] = &DeclarativeSubscription{
+		Comp: comp,
+		Subscription: rtpubsub.Subscription{
+			PubsubName:      comp.Spec.Pubsubname,
+			Topic:           comp.Spec.Topic,
+			DeadLetterTopic: comp.Spec.DeadLetterTopic,
+			Metadata:        comp.Spec.Metadata,
+			Rules:           []*rtpubsub.Rule{{Path: "/"}},
+		},
+	}
 }
 
-func (c *ComponentStore) ListSubscriptions() []rtpubsub.Subscription {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-	return c.subscriptions
+func (c *ComponentStore) DeleteStreamSubscription(names ...string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	for _, name := range names {
+		delete(c.subscriptions.streams, name)
+	}
 }
 
-func (c *ComponentStore) AddDeclarativeSubscription(subscriptions ...subapi.Subscription) error {
+func (c *ComponentStore) DeleteDeclarativeSubscription(names ...string) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	for _, sub := range subscriptions {
-		for _, existing := range c.declarativeSubscriptions {
-			if existing.ObjectMeta.Name == sub.ObjectMeta.Name {
-				return fmt.Errorf("subscription with name %s already exists", sub.ObjectMeta.Name)
+	for _, name := range names {
+		delete(c.subscriptions.declaratives, name)
+		for i, existing := range c.subscriptions.declarativesList {
+			if existing == name {
+				c.subscriptions.declarativesList = append(c.subscriptions.declarativesList[:i], c.subscriptions.declarativesList[i+1:]...)
+				break
 			}
 		}
 	}
+}
+
+func (c *ComponentStore) ListSubscriptions() []rtpubsub.Subscription {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
 
-	c.declarativeSubscriptions = append(c.declarativeSubscriptions, subscriptions...)
+	var subs []rtpubsub.Subscription
+	taken := make(map[string]int)
 
-	return nil
+	for _, name := range c.subscriptions.declarativesList {
+		sub := c.subscriptions.declaratives[name].Subscription
+		key := sub.PubsubName + "||" + sub.Topic
```

Comment:
We have the concept of namespaced pub/sub - does this require a change here? https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-namespace/

Reply:
To me- it looks like this namespacing is done on the PubSub component spec level. The _name_ of the pubsub still needs to be unique for that namespace, so there would be no possible for conflict here.